### PR TITLE
Change interface of add_source_parser() and add add_source_suffix()

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Refactoring
- Bugfix (in master branch: see #4564)

### Purpose
- Separate registering of suffix and parser

### Detail
- Please read the diff of `doc/extdev/appapi.rst`.

### CODE REVIEW

File: doc/extdev/parserapi.rst [hunk 1]
```diff
@@ -3,6 +3,35 @@
...
+    The Parser analyzes the the input document and creates a node tree
+    representation.
+
+__ http://docutils.sourceforge.net/docs/dev/hacking.html#parsing-the-document
+
+In Sphinx, the parser modules works as same as docutils.  The parsers are
+registered to Sphinx by extensions using Application APIs;
+:meth:`Sphinx.add_source_suffix()` and :meth:`Sphinx.add_source_parsers()`.
+
+The source_suffix is a mapping from file suffix to file type.  For example,
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: s/source_suffix/:confval:`source_suffix`/ ?

File: doc/extdev/parserapi.rst [hunk 2]
```diff
@@ -3,6 +3,35 @@
...
+``.rst`` file is mapped to ``'restrcturedtext'`` type.  Sphinx uses the
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: s/restrcturedtext/restructuredtext/

File: doc/extdev/parserapi.rst [hunk 3]
```diff
@@ -3,6 +3,35 @@
...
+The users can override the source_suffix mappings using
+:confval:`source_suffix` like following::
+
+    # a mapping from file suffix to file types
+    source_suffix = {
+        '.rst': 'restrcturedtext',
+        '.md': 'markdown',
+    }
+
+So please let users know what file types your parser supports.  Then users
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: This paragraph doesn't make sense. I _think_ you mean:

> You will need to configure this setting in order to use a different file types for a given parser.

File: sphinx/registry.py [hunk 4]
```diff
@@ -239,10 +239,23 @@ def add_source_suffix(self, suffix, filetype):
         else:
             self.source_suffix[suffix] = filetype
 
-    def add_source_parser(self, suffix, parser):
-        # type: (unicode, Type[Parser]) -> None
-        logger.debug('[app] adding search source_parser: %r, %r', suffix, parser)
-        self.add_source_suffix(suffix, suffix)
+    def add_source_parser(self, *args):
```
Comment (stephenfin) [inline sphinx/registry.py:242]: nit: How about doing this instead?

    empty_parser = object()

    def add_source_parser(self, suffix, parser=empty_parser):
        if parser != empty_parser:
            # emit warning here

It would keep a similarly (clear) function style and allow you to do the same thing. It works because every `object` instance is unique. You could also use `parser=None`, but if the user called this with `None` intentionally, they wouldn't see the issue

**Review by stephenfin — CHANGES_REQUESTED** [review]
Couple of typos plus some helpful suggestions

File: sphinx/registry.py (hunk 4 above)
Comment (tk0miya) [reply sphinx/registry.py:242]: Do you mean like this? I feel `parser = suffix` is strange.
```
def add_source_parser(self, suffix, parser=empty_parser):
    if parser != empty_parser:
        # old style arguments: (suffix, source_parser)
        warnings.warn('app.add_source_parser() does not support suffix argument. '
                      'Use app.add_source_suffix() instead.',
                      RemovedInSphinx30Warning)
    else:
        # new sytle arguments: (source_parser)
        suffix = None
        parser = suffix
```

>You could also use parser=None, but if the user called this with None intentionally, they wouldn't see the issue

I don't mind. If developers gives such value, Sphinx will be crashed (without this change).

File: doc/extdev/parserapi.rst (hunk 3 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: I meant that the author of Parser should tell a file type supported by the Parser to users.
The users need to know it to configure `source_suffix` variable.

I think the target of this document is a third party developer. So your paragraph is not good for them.

File: doc/extdev/parserapi.rst (hunk 1 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: No, this `source_suffix` is a concept. Both users and extensions can modify it using `:confval:source_suffix` and `add_source_suffix()` API. I'd like to describe it here because the authors of parser module should understand how it works.

File: doc/extdev/parserapi.rst (hunk 2 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: done

**Comment by tk0miya** [comment]
Thank you for comment!

File: doc/extdev/parserapi.rst (hunk 1 above)
Comment (stephenfin) [reply doc/extdev/parserapi.rst]: That's fair. I'd probably remove the underscore in that case and make it italic so it stands out:

    *source suffix*

Otherwise though, this is fine

File: doc/extdev/parserapi.rst (hunk 3 above)
Comment (stephenfin) [reply doc/extdev/parserapi.rst]: Oh, OK. I understand now. This could still do with clarification though, given that it's already confused one person (me!). How about:

    You should indicate file types your parser supports. This will allow users to configure their settings appropriately.

Or something like that

File: sphinx/registry.py (hunk 4 above)
Comment (stephenfin) [reply sphinx/registry.py:242]: Ohhh, wait a second. I mixed up the order of the parameters. `suffix` is the first parameter - not the second. You can ignore this comment so 😄

File: doc/extdev/parserapi.rst (hunk 1 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: Sounds good.

File: doc/extdev/parserapi.rst (hunk 3 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: Very good, thanks!

File: sphinx/registry.py (hunk 4 above)
Comment (tk0miya) [reply sphinx/registry.py:242]: Okay :-)

**Review by tk0miya — COMMENTED** [review]
Updated.

File: doc/extdev/parserapi.rst [hunk 5]
```diff
@@ -3,6 +3,35 @@
...
+file type to looking for parsers from registered list.  On searching,
+Sphinx refers ``Parser.supported`` attribute, and picks up a parser which
+contains the file type in the attribute.
+
+The users can override the source suffix mappings using
+:confval:`source_suffix` like following::
+
+    # a mapping from file suffix to file types
+    source_suffix = {
+        '.rst': 'restrcturedtext',
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: Oops. I assume you meant `restructuredtext`?

File: doc/extdev/parserapi.rst [hunk 6]
```diff
@@ -3,6 +3,35 @@
 Parser API
 ==========
 
+`The document of docutils describes`__ about the parsers like following:
+
+    The Parser analyzes the the input document and creates a node tree
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: s/the the/the/

File: doc/extdev/parserapi.rst [hunk 7]
```diff
@@ -3,6 +3,35 @@
 Parser API
 ==========
 
+`The document of docutils describes`__ about the parsers like following:
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: The docutils documentation describes parsers as follows:

File: doc/extdev/parserapi.rst [hunk 8]
```diff
@@ -3,6 +3,35 @@
...
+__ http://docutils.sourceforge.net/docs/dev/hacking.html#parsing-the-document
+
+In Sphinx, the parser modules works as same as docutils.  The parsers are
+registered to Sphinx by extensions using Application APIs;
+:meth:`Sphinx.add_source_suffix()` and :meth:`Sphinx.add_source_parsers()`.
+
+The *source suffix* is a mapping from file suffix to file type.  For example,
+``.rst`` file is mapped to ``'restructuredtext'`` type.  Sphinx uses the
+file type to looking for parsers from registered list.  On searching,
+Sphinx refers ``Parser.supported`` attribute, and picks up a parser which
```
Comment (stephenfin) [inline doc/extdev/parserapi.rst]: refers to the ``Parser.supported`` attribute and picks

(missing 'to the'. Drop the comma before 'and')

**Review by stephenfin — COMMENTED** [review]
I missed a few things 😞

File: doc/extdev/parserapi.rst (hunk 7 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: done

File: doc/extdev/parserapi.rst (hunk 6 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: It seems this is a typo of original text.
https://github.com/docutils-mirror/docutils/blob/master/docs/dev/hacking.txt#L69

File: doc/extdev/parserapi.rst (hunk 8 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: done

File: doc/extdev/parserapi.rst (hunk 5 above)
Comment (tk0miya) [reply doc/extdev/parserapi.rst]: done

**Review by tk0miya — COMMENTED** [review]
Fixed again

File: sphinx/environment/__init__.py [hunk 9]
```diff
@@ -481,7 +482,7 @@ def get_outdated_files(self, config_changed):
                     changed.add(docname)
                     continue
                 # finally, check the mtime of dependencies
-                for dep in self.dependencies[docname]:
+                for dep in self.dependencies[None].union(self.dependencies[docname]):
```
Comment (tk0miya) [inline sphinx/environment/__init__.py]: oh, I noticed this change is not related with this PR.
I have to remove this.

**Review by tk0miya — COMMENTED** [review]
I will merge this soon after removal.

**Comment by tk0miya** [comment]
Merged.
@stephenfin I'm merging this now to step forward. Please let me know if you still have any opinion for this.

**Comment by stephenfin** [comment]
Yup, I was on vacation. This looks good to me. Thanks for the follow-ups

**Comment by tk0miya** [comment]
Have a good vacation!
